### ISSUE TITLE
clarify documentation on gplazma.authz.upload-directory

### ISSUE THREAD

**Issue by calestyo** [dCache/dcache#3300]
Hi.

The documentation of gplazma.authz.upload-directory says:
"Users are restricted so they may only access files beneath their root directory. The upload directory is an exception to this rule."

AFAIK, the was *not* the case in previous (< 2.16) versions of dCache, or at least I was told so I think...
So I had e.g.:
```
/etc/dcache/$ grep -Ri "\.upload-directory"
dcache.conf:webdav.authz.allowed-paths=${somepaths}:${webdav.authz.upload-directory}:/robots.txt
dcache.conf:#xrootd.authz.write-paths=${somepaths}:${xrootd.authz.upload-directory}
```

(For webdav there doesn't seem to be a property to control read/write access independently, so I used allowed-paths, which seems to control both at once?)

So EITHER, that gplazma.authz.upload-directory is always implicitly allowed) isn't the case...
OR it should be documented at least in the respective options (xrootd.authz.write-paths, etc. pp.) that a further dir is implicitly added... AND also in the release notes, if this was actually a change in behaviour.

Cheers,
Chris.

**Comment by calestyo** [comment]
Also the part "Users are restricted so they may only access files beneath their root directory." seems ambiguous:

If I'd change the root of a door from / to say /foo... how could one ever access /upload (regardless of any permissions)... ?
The respective door would use /foo as /... so it don't quite understand how this can have to do anything with the namespace root (e.g. what doors default to =dcache.root).

**Comment by paulmillar** [comment]
Taking the first comment first: xrootd, ftp, and webdav services had properties (`xrootd.authz.upload-directory`, `ftp.authz.upload-directory`, and `webdav.authz.upload-directory` respectively) that allowed users to access certain paths outside of their user-root (NB. this is *user* root, not door root).  These all took `dcache.upload-directory` as a default value, which itself took `/upload` as the default value.  Again, note that this is *only* about bypassing the user's (personal) root directory; if the user has a root of `/` then these properties had no effect.

With 2.16, the user-root behaviour was refactored, with a single place for configuring this behaviour.  Nothing has (or nothing should have) changed in how dCache behaves, but way in which dCache is configured has changed.

The properties you cited (e.g., `wedav.authz.allowed-paths`) is about authorising read and/or write activity.  These door-specific restrictions (if any) are applied in addition to a restriction preventing the user from accessing outside of the user's root directory.

With your second comment: it seems you misunderstand what is a user's root.  This is a property of the user's login session that is completely independent of which protocol and which door that user authenticated.  A non-`/` value for the user root means that this particular user only sees a subset of dCache namespace.

**Comment by calestyo** [comment]
Ah, I see,... so this neither the "real" root, nor the home dir of a user, but the user's visible part of the name space.
E.g. if the global root is set to / and the user-root to /pnfs and his home to /pnfs/foo and there is in addition /secret,… he'd see /pnfs (and implicitly /upload), but not /secret.
Right?

Then still, the:
 "Users are restricted so they may only access files beneath their root directory. The upload directory is an exception to this rule."
should IMO be improved, e.g. to:
 "Users are restricted so they may only access files beneath their per-user-root directory (which is not the general root directory of the respective door). The upload directory is an exception to this rule."

In addition, I'd suggest that one adds to the documentation of the dcache.upload-directory and gplazma.authz.upload-directory ... that these directories should likely be included in wedav.authz.allowed-paths and xrootd.authz.write-paths

:-)

I'd have made pull request: #3315

**Comment by paulmillar** [comment]
No.  If a user with root `/foo` and dCache has directory `/foo/bar` then the user will see `bar` if they list the contents of the root directory.

**Comment by calestyo** [comment]
Uhm... o.O

But then again,.. how could it ever access /upload?
If there is `/foo/bar` as well as `/upload`  and the user has root `/foo`, and he lists `/` and sees `bar`, he should never be able to access `/upload` (or is this just done internally by dCache and not the client?), cause if he could, there would be an ambiguity between
`/upload` and `/foo/upload`

**Comment by paulmillar** [comment]
READ THE DOCUMENTATION (dcache.upload-directory)

This is clearly explained as a problem, with three possible solutions.
